Replace TemporaryFile() with io.BytesIO()

Ankit-Dhankhar opened the pull request:
Using similar syntax as torch_tensor_serializer/desearializer. #2357

midokura-silvia commented:
Ok, I just confirmed that there is a unit test. Could we improve the unit test by using `@pytest.mark.parametrize` and call it for let's say 4 different arrays? Some combination of float and int arrays, with non-symmetric dimensions? The current choice of `numpy.ones((10, 10))` feels a bit limited.

Ankit-Dhankhar commented:
Sure I will add test cases.

Ankit-Dhankhar commented:
I have added test cases for random integer arrays of non-symmetric dimensions.
But for array with float value it is giving `assert tensor_serialized[0] != serde.NO_COMPRESSION` this false.

Ankit-Dhankhar commented on syft/serde/torch_serde.py:
```diff
@@ -59,30 +58,38 @@ def _deserialize_tensor(tensor_bin) -> torch.Tensor:
             "Detaching tensor to continue"
         )
-        tensor = tensor.detach()
+        tensor = torch.detach()
```
It is a typo mistake. As there was no test case for "requires_grad=True", I missed it.

Ankit-Dhankhar replied:
Updated

Ankit-Dhankhar commented on test/test_serde.py:
```diff
@@ -568,14 +568,20 @@ def test_pointer_tensor_detail(id):
+@pytest.mark.parametrize(
+    "tensor",
+    [
+        (torch.tensor(numpy.ones((10, 10)), requires_grad=False)),
```
Yes, I also believe that we need it else mistake like pointed in the previous comment will not be detected.

Ankit-Dhankhar replied:
Added test case

Ankit-Dhankhar commented on syft/serde/torch_serde.py:
```diff
@@ -59,30 +58,38 @@ def _deserialize_tensor(tensor_bin) -> torch.Tensor:
 def numpy_tensor_serializer(tensor: torch.Tensor) -> bin:
     """Strategy to serialize a tensor using numpy npy format.
-    If tensor requires to calculate gradients, it will be detached.
+    If tensor requires to calculate gradients, it will detached.
```
Updated

Ankit-Dhankhar commented:
I have added test-cases with requires_grad=True but tensor generated with torch.randn([dim]) fails this. Is this expected?

> I have added test cases for random integer arrays of non-symmetric dimensions.
> But for array with float value it is giving `assert tensor_serialized[0] != serde.NO_COMPRESSION` this false.

midokura-silvia commented:
@robert-wagner any update from your side?

iamtrask commented:
@midokura-silvia - if it's been ~7 days or more after checking in with someone who reviewed - and you feel the PR should be approved - you may go ahead and merge.

iamtrask commented:
(in this case it's been 3 months - more than enough time)

gmuraru commented:
Can I pick this up?

karlhigley commented:
@Ankit-Dhankhar I cloned your fork, squashed the commits in this PR, and rebased it on current PySyft master. You can see the [resulting diff](https://github.com/OpenMined/PySyft/commit/39b7f23a25834299f99d6bee062efee0002499d2) over here on the `pr/2376` branch.

Sorry it's taken so long! Okay to force-push the result to your branch/PR so we can finally get this merged? 😄
